### PR TITLE
fix(NODE-5840): heartbeat duration includes socket creation

### PR DESCRIPTION
**Pull request by nbbeeken**
#### What is changing?

- Refactor connect to be async
- Rename and refactor makeConnection -> makeSocket
- Add makeConnection function
- Refactor monitor to record start time after socket is created, before handshake
- socket failures set heartbeat failed duration to -1 
- Refactor tests using callback style connect
- Test duration does not include socket creation by artificially delaying it

#### What is the motivation for this change?

The duration field of hb started and failed should match the definition in the spec.

### Fixed heartbeat duration including socket creation

The ServerHeartbeatSucceeded and ServerHeartbeatFailed event have a duration property that represents the time it took to perform the `hello` handshake with MongoDB. The Monitor responsible for issuing heartbeats mistakenly included the time it took to create the socket in this field, which inflates the value with the time it takes to perform a DNS lookup, TCP, and TLS handshakes.

### Double-check the following

- [x] Ran `npm run check:lint` script
- [x] Self-review completed using the [steps outlined here](https://github.com/mongodb/node-mongodb-native/blob/HEAD/CONTRIBUTING.md#reviewer-guidelines)
- [x] PR title follows the [correct format](https://www.conventionalcommits.org/en/v1.0.0/): `type(NODE-xxxx)[!]: description`
  - Example: `feat(NODE-1234)!: rewriting everything in coffeescript`
- [x] Changes are covered by tests
- [x] New TODOs have a related JIRA ticket

### CODE REVIEW

File: src/cmap/connect.ts [hunk 1]
```diff
@@ -435,78 +414,71 @@ function loadSocks() {
...
+      destination: {
+        host: destination.host,
+        port: destination.port
+      },
+      proxy: {
+        // host and port are ignored because we pass existing_socket
+        host: 'iLoveJavaScript',
+        port: 0,
+        type: 5,
+        userId: options.proxyUsername || undefined,
```
Comment (alenakhineika) [inline src/cmap/connect.ts:453]: Is setting `undefined` explicitly necessary? If `proxyUsername` or `proxyPassword` are not in the options, their value will be `undefined` anyway.

File: src/error.ts [hunk 2]
```diff
@@ -150,7 +150,7 @@ export class MongoError extends Error {
   }
 
   /** @internal */
-  static buildErrorMessage(e: Error | string): string {
+  static buildErrorMessage(e: unknown): string {
```
Comment (alenakhineika) [inline src/error.ts:153]: What is the motivation for this change? I don't see where in the code it could be set differently rather than `Error` or `string`.

File: src/sdam/monitor.ts [hunk 3]
```diff
@@ -352,37 +354,45 @@ function checkServer(monitor: Monitor, callback: Callback<Document | null>) {
...
-    if (err) {
-      monitor.connection = null;
-
-      awaited = false;
-      onHeartbeatFailed(err);
-      return;
+  (async () => {
+    const socket = await makeSocket(monitor.connectOptions);
+    const connection = makeConnection(monitor.connectOptions, socket);
+    start = now();
```
Comment (alenakhineika) [inline src/sdam/monitor.ts:360]: Maybe it would be nice to add a comment here and on the 338 line (with another `start = now()`) explaining what you were telling me about when we expect it to be called.

File: src/utils.ts [hunk 4]
```diff
@@ -492,9 +492,9 @@ export function now(): number {
 }
 
 /** @internal */
-export function calculateDurationInMs(started: number): number {
+export function calculateDurationInMs(started: number | undefined): number {
```
Comment (alenakhineika) [inline src/utils.ts:495]: I also didn't find where started could be set as `undefined`.

File: test/integration/connection-monitoring-and-pooling/connection.test.ts [hunk 5]
```diff
@@ -35,15 +33,15 @@ describe('Connection', function () {
     it('should execute a command against a server', {
       metadata: { requires: { apiVersion: false, topology: '!load-balanced' } },
       test: async function () {
-        const connectOptions: Partial<ConnectionOptions> = {
+        const connectOptions: ConnectionOptions = {
```
Comment (alenakhineika) [inline test/integration/connection-monitoring-and-pooling/connection.test.ts:36]: Why do we remove `Partial` here? To get rid of those castings later `connectOptions as any as ConnectionOptions`? Because connectOptions has indeed partial type here.

File: test/unit/sdam/monitor.test.ts [hunk 6]
```diff
@@ -579,4 +582,32 @@ describe('monitoring', function () {
...
+        socket.on('data', () => socket.destroy(new Error('I am not real!')));
+        return socket;
+      });
+
+      client = new MongoClient(`mongodb://localhost:1`, { serverSelectionTimeoutMS: 200 });
+      client.on('serverHeartbeatFailed', ev => (serverHeartbeatFailed = ev));
+    });
+
+    afterEach(function () {
+      sinon.restore();
```
Comment (alenakhineika) [inline test/unit/sdam/monitor.test.ts:604]: Do we need to destroy the socket in afterEach? In case something went wrong and we haven't destroyed it on data callback.

File: src/cmap/connect.ts (hunk 1 above)
Comment (nbbeeken) [reply src/cmap/connect.ts:453]: If proxyUsername is an empty string we pass in `undefined` instead

File: src/error.ts (hunk 2 above)
Comment (nbbeeken) [reply src/error.ts:153]: I made the type in `connectionFailureError` more accurate which makes error nullish. To me it felt easier to propagate the handling of that to `buildErrorMessage`

Comment (alenakhineika) [reply src/error.ts:153]: Isn't it then `Error | string | null`?

File: test/integration/connection-monitoring-and-pooling/connection.test.ts (hunk 5 above)
Comment (nbbeeken) [reply test/integration/connection-monitoring-and-pooling/connection.test.ts:47]: I rounded out the type so its no longer partial :)

Rather than cast away a problem we should mark why we are able to get away with violating TS but converting tests to TS is sometimes done to large files and we can leave behind type errors that don't actually break the code at runtime.

File: src/utils.ts (hunk 4 above)
Comment (nbbeeken) [reply src/utils.ts:495]: Isn't it fine for a let to remain unassigned? typescript will know if you use it before assignment right?

File: src/sdam/monitor.ts (hunk 3 above)
Comment (nbbeeken) [reply src/sdam/monitor.ts:364]: Added doc comments

File: src/error.ts (hunk 2 above)
Comment (nbbeeken) [reply src/error.ts:153]: Any error caught in a catch block has the potential to be any type, we just disabled that rule temporarily, `useUnknownInCatchVariables`, I'm writing unknown here now so there is less work for NODE-3659 down the road.

**Review by alenakhineika — APPROVED** [review]

**Review by W-A-James — APPROVED** [review]
